shared/actions/login/saga.js

```diff
@@ -0,0 +1,588 @@
+    console.log(error)
+    yield put(Creators.loginDone(error))
+  } else {
+    yield put(Creators.loginDone())
+  }
+  yield call(navBasedOnLoginState)
+}
+
+function * loginFlowSaga (usernameOrEmail) {
+  const loginSagas = kex2Sagas(cancelLogin, provisionerSuccessInLoginSaga, finishLoginSaga)
+  const catchError = function * () { yield call(cancelLogin) }
+
+  const channelConfig = Saga.singleFixedChannelConfig(Object.keys(loginSagas))
+  const loginChanMap = yield call(loginRpc, channelConfig, usernameOrEmail)
+
+  // Returns an Array<Task>
+  // If there are any unexpected errors let's cancel the login. The error will be shown as a global error
+  yield _.map(loginSagas, (saga, methodName) => Saga.effectOnChannelMap(c => Saga.safeTakeLatestWithCatch(c, catchError, saga), loginChanMap, methodName))
+}
+
+function * initalizeMyCodeStateForLogin () {
+  // We can either be a newDevice or an existingDevice. Here in the login
+  // flow, let's set ourselves to be a newDevice
+  yield put(Creators.setMyDeviceCodeState(
+    isMobile ? Constants.codePageDeviceRoleNewPhone : Constants.codePageDeviceRoleNewComputer,
+  ))
+}
+
+function * initalizeMyCodeStateForAddingADevice () {
+  // We can either be a newDevice or an existingDevice. Here in the adding a device
+  // flow, let's set ourselves to be an existing device
+  yield put(Creators.setMyDeviceCodeState(
+    isMobile ? Constants.codePageDeviceRoleExistingPhone : Constants.codePageDeviceRoleExistingComputer,
+  ))
+}
+
+function * startLoginSaga () {
+  yield put(navigateAppend(['usernameOrEmail'], [loginTab, 'login']))
+
+  yield call(initalizeMyCodeStateForLogin)
+
+  const {onBack, onSubmit} = yield race({
+    onBack: take(Constants.onBack),
+    onSubmit: take(Constants.submitUsernameOrEmail),
+  })
+
+  if (onBack) {
+    yield call(cancelLogin)
+  } else if (onSubmit) {
+    const usernameOrEmail = onSubmit.payload.usernameOrEmail
+    yield call(loginFlowSaga, usernameOrEmail)
+  }
+}
+
+function * cameraBrokenModeSaga ({payload: {broken}}) {
+  const codePage: AfterSelect<typeof codePageSelector> = yield select(codePageSelector)
+  if (codePage.myDeviceRole == null) {
+    console.warn("my device role is null, can't setCameraBrokenMode. Bailing")
+    return
+  }
+
+  if (codePage.otherDeviceRole == null) {
+    console.warn("other device role is null, can't setCameraBrokenMode. Bailing")
+    return
+  }
+
+  const mode = defaultModeForDeviceRoles(codePage.myDeviceRole, codePage.otherDeviceRole, broken)
+  if (!mode) {
+    console.warn("mode is null!, can't setCodePageMode. Bailing")
+    return
+  }
+  yield put(Creators.setCodePageMode(mode))
+}
+
+function * loginSuccess () {
+  yield put(Creators.loginDone())
+  yield put(loadDevices())
+  yield put(bootstrap())
+}
+
+function * addNewDeviceSaga ({payload: {role}}: DeviceConstants.AddNewDevice) {
+  yield call(initalizeMyCodeStateForAddingADevice)
+
+  const onBackSaga = function * (response) {
+    yield put(loadDevices())
```

Comment:
:speech_balloon: Bummer that we have to explicitly call this here and in other places. We should move this into the connected component init or some kind of route entry handler.